osu.Game/Overlays/Mods/ModSelectOverlay.cs

```diff
@@ -313,22 +321,29 @@ protected override void LoadComplete()
             if (AllowCustomisation)
                 ((IBindable<IReadOnlyList<Mod>>)modSettingsArea.SelectedMods).BindTo(SelectedMods);
 
-            SelectedMods.BindValueChanged(_ =>
+            SelectedMods.BindValueChanged(mods =>
             {
-                updateRankingInformation();
+                var newMods = ActiveMods.Value.Except(mods.OldValue).Concat(mods.NewValue).ToList();
```

Comment:
I don't like this...

The mod select overlay is (unfortunately) very particular about what mod instances it's operating on. Incorrectly handling this will lead to failures like something not updating correctly after an external set of `SelectedMods`.

_At the least_ this should be done after `updateFromExternalSelection()` is done. But... ideally I'd like to not have to play these replace games. I'd rather have something along the lines of (warning, untested):

```diff
diff --git a/osu.Game/Overlays/Mods/ModSelectOverlay.cs b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
index 6b80ff6e44..dff03597be 100644
--- a/osu.Game/Overlays/Mods/ModSelectOverlay.cs
+++ b/osu.Game/Overlays/Mods/ModSelectOverlay.cs
@@ -103,6 +103,7 @@ public string SearchTerm
         protected virtual ModColumn CreateModColumn(ModType modType) => new ModColumn(modType, false);
 
         protected virtual IReadOnlyList<Mod> ComputeNewModsFromSelection(IReadOnlyList<Mod> oldSelection, IReadOnlyList<Mod> newSelection) => newSelection;
+        protected virtual IReadOnlyList<Mod> ComputeActiveMods(IReadOnlyList<Mod> selectedMods) => selectedMods;
 
         protected virtual IEnumerable<ShearedButton> CreateFooterButtons()
         {
@@ -323,11 +324,9 @@ protected override void LoadComplete()
 
             SelectedMods.BindValueChanged(mods =>
             {
-                var newMods = ActiveMods.Value.Except(mods.OldValue).Concat(mods.NewValue).ToList();
-                ActiveMods.Value = newMods;
-
                 updateFromExternalSelection();
                 updateCustomisation();
+                ActiveMods.Value = ComputeActiveMods(SelectedMods.Value);
             }, true);
 
             ActiveMods.BindValueChanged(_ =>
diff --git a/osu.Game/Screens/OnlinePlay/Match/RoomModSelectOverlay.cs b/osu.Game/Screens/OnlinePlay/Match/RoomModSelectOverlay.cs
index d3fd6de911..466398f61a 100644
--- a/osu.Game/Screens/OnlinePlay/Match/RoomModSelectOverlay.cs
+++ b/osu.Game/Screens/OnlinePlay/Match/RoomModSelectOverlay.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.Linq;
 using osu.Framework.Allocation;
@@ -9,6 +10,7 @@
 using osu.Game.Overlays;
 using osu.Game.Overlays.Mods;
 using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
 
 namespace osu.Game.Screens.OnlinePlay.Match
 {
@@ -29,17 +31,19 @@ protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            selectedItem.BindValueChanged(v =>
+            selectedItem.BindValueChanged(v => ComputeActiveMods(SelectedMods.Value), true);
+        }
+
+        protected override IReadOnlyList<Mod> ComputeActiveMods(IReadOnlyList<Mod> selectedMods)
+        {
+            if (selectedItem.Value is PlaylistItem item)
             {
-                if (v.NewValue is PlaylistItem item)
-                {
-                    var rulesetInstance = rulesets.GetRuleset(item.RulesetID)?.CreateInstance();
-                    Debug.Assert(rulesetInstance != null);
-                    ActiveMods.Value = item.RequiredMods.Select(m => m.ToMod(rulesetInstance)).Concat(SelectedMods.Value).ToList();
-                }
-                else
-                    ActiveMods.Value = SelectedMods.Value;
-            }, true);
+                var rulesetInstance = rulesets.GetRuleset(item.RulesetID)?.CreateInstance();
+                Debug.Assert(rulesetInstance != null);
+                return item.RequiredMods.Select(m => m.ToMod(rulesetInstance)).Concat(selectedMods).ToList();
+            }
+
+            return selectedMods;
         }
     }
 }

```

Is this feasible?

Reply:
Sounds good.